sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,530 @@
+            CipherFactory.initCipher(clone.cipher, directory, Optional.of(cipher.getIV()), Cipher.DECRYPT_MODE, getFilePointer() + off);
+            clone.tmpBuffer = ByteBuffer.allocate(CHUNK_SIZE);
+            return clone;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
+            if (offset < 0 || length < 0 || offset + length > this.length()) {
+                throw new IllegalArgumentException(
+                    "slice() "
+                        + sliceDescription
+                        + " out of bounds: offset="
+                        + offset
+                        + ",length="
+                        + length
+                        + ",fileLength="
+                        + this.length()
+                        + ": "
+                        + this
+                );
+            }
+            return new CryptoBufferedIndexInput(
+                getFullSliceDescription(sliceDescription),
+                channel,
+                off + offset,
+                length,
+                getBufferSize(),
+                cipher,
+                directory
+            );
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public final long length() {
+            return end - off;
+        }
+
+        private int read(ByteBuffer dst, long position) throws IOException {
+            int ret;
+            int i;
+            tmpBuffer.rewind();
+            // FileChannel#read is forbidden
+            synchronized (channel) {
+                channel.position(position);
+                i = stream.read(tmpBuffer.array(), 0, dst.remaining());
+            }
+            tmpBuffer.limit(i);
+            try {
+                if (end - position > i) ret = cipher.update(tmpBuffer, dst);
+                else ret = cipher.doFinal(tmpBuffer, dst);
+            } catch (ShortBufferException | IllegalBlockSizeException | BadPaddingException ex) {
+                throw new IOException("failed to decrypt blck.", ex);
+            }
+            return ret;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected void readInternal(ByteBuffer b) throws IOException {
+            long pos = getFilePointer() + off;
+
+            if (pos + b.remaining() > end) {
+                throw new EOFException(
+                    Thread.currentThread().getId()
+                        + " read past EOF: "
+                        + this
+                        + " isClone? "
+                        + isClone
+                        + " off: "
+                        + off
+                        + " pos: "
+                        + pos
+                        + " end: "
+                        + end
+                );
+            }
+
+            try {
```

Comment:
Can this be revised to be a little bit easier to follow. I am sure it works great, but it is just a little tricky to read with the simplified temp vars.